### PR TITLE
Backend

### PR DESCRIPTION
**Pull request by nfischer**
This is to merge in most of the basic backend functionality. Since this is a big change, I'm making this a PR that someone on frontend team and someone on backend should review.

While this is in review, I'll be continuing to add functionality to the backend. If there's anything you need, please leave it on here as a comment.

### CODE REVIEW

**Comment by nfischer** [comment]
I added in the exception type

**Comment by sakekasi** [comment]
looks good to me. will approve once the issue where children are deleted is resolved

**Comment by nfischer** [comment]
I'll try working on that after class. If anyone wants to tackle it, it's a "TODO" comment

File: backend/cdApp.py [hunk 1]
```diff
@@ -37,35 +37,138 @@ def dict_factory(cursor, row):
...
+    #         children[child['name']] = child['dest']
+    #     ret_node['children'] = children
+    #     cursor.close()
+    #     return ret_node
+
+
+    # /nodes/tree
+    # returns the tree of nodes
+    # rootId is hard coded right now. Need clarification on that
+    def get(selft):
```
Comment (sakekasi) [inline backend/cdApp.py]: misspelled self

File: backend/cdApp.py [hunk 2]
```diff
@@ -37,35 +37,138 @@ def dict_factory(cursor, row):
...
+                              FROM nodes AS n LEFT JOIN children AS c 
+                              ON n.id = c.parent_id WHERE id=(?)''', node_id)
+        return cursor.fetchone()
+
+class Children(Resource):
+    def put(self):
+        g.db.execute('INSERT INTO children (parent_id, children) values (?, ?)',
+                     [request.form['id'], request.form['children']])
+        g.db.commit()
+        return jsonify(message='Children were successfully added to the node', id='1')
```
Comment (sakekasi) [inline backend/cdApp.py]: shouldn't Id be a variable, not a constant '1'?

File: backend/cdApp.py [hunk 3]
```diff
@@ -37,35 +37,138 @@ def dict_factory(cursor, row):
...
-        g.db.execute('insert into nodes (contents, renderer) values (?, ?)',
+        """
+        This inserts a new node into the database, with the node_id
+        automatically assigned
+        """
+        g.db.execute('INSERT INTO nodes (contents, renderer) VALUES(?, ?)',
                      [request.form['contents'], request.form['renderer']])
         g.db.commit()
-        return jsonify(message='New node was successfully created')
+        return jsonify(message='New node was successfully created', id='1')
```
Comment (sakekasi) [inline backend/cdApp.py]: need actual node Id here

File: backend/cdApp.py [hunk 4]
```diff
@@ -37,35 +37,138 @@ def dict_factory(cursor, row):
...
-            # g.db.commit()
-            return "FOO" #jsonify(message='New node was successfully created')
+    def post(self, node_id):
+        g.db.execute('''UPDATE nodes 
+                        SET contents=(?),renderer=(?) 
+                        WHERE id=(?)''',
+                        [request.form['contents'], request.form['renderer'], node_id])
+        g.db.commit()
+        g.db.commit()
+        return jsonify(message='Node was successfully updated.', id='1')
```
Comment (sakekasi) [inline backend/cdApp.py]: need actual node id here

File: backend/cdApp.py (hunk 1 above)
Comment (nfischer) [inline backend/cdApp.py]: I'll tackle this

File: backend/cdApp.py [hunk 5]
```diff
@@ -37,35 +37,138 @@ def dict_factory(cursor, row):
...
+
+
+    # /nodes/tree
+    # returns the tree of nodes
+    # rootId is hard coded right now. Need clarification on that
+    def get(selft):
+        cursor = g.db.execute('''SELECT n.id, n.contents, n.renderer, c.children 
+                              FROM nodes AS n LEFT JOIN children AS c 
+                              ON n.id = c.parent_id''');
+        rv = {}
```
Comment (nfischer) [inline backend/cdApp.py]: @Sarabostani What does "rv" represent? Is this the entire tree?

Comment (sakekasi) [inline backend/cdApp.py]: i think it's "return value." In this case, I think that the actual behavior does the right thing, but I agree that rv isn't the best name

File: backend/cdApp.py [hunk 6]
```diff
@@ -37,35 +37,138 @@ def dict_factory(cursor, row):
...
-        rv['children'] = children
-        cur.close()
-        return rv
+        cursor = g.db.execute('''SELECT n.id, n.contents, n.renderer, c.children 
+                              FROM nodes AS n LEFT JOIN children AS c 
+                              ON n.id = c.parent_id WHERE id=(?)''', node_id)
+        return cursor.fetchone()
+
+class Children(Resource):
+    def put(self):
```
Comment (sakekasi) [inline backend/cdApp.py]: `put(self, node_id):`

File: backend/cdApp.py (hunk 5 above)
Comment (Sarabostani) [inline backend/cdApp.py]: yes, since tree is a dictionary not an array.

File: backend/cdApp.py (hunk 2 above)
Comment (sakekasi) [inline backend/cdApp.py]: `return jsonify(message='Children ...', id=node_id)`

File: backend/cdApp.py (hunk 5 above)
Comment (sakekasi) [inline backend/cdApp.py]: is tree a dict? I was expecting a root id and an array of nodes

File: backend/cdApp.py (hunk 3 above)
Comment (sakekasi) [inline backend/cdApp.py]: looks like this is the non-trivial one. would need to select the node and return its id? or does the id get returned somehow form the execution of the insert statement?

File: backend/cdApp.py (hunk 4 above)
Comment (sakekasi) [inline backend/cdApp.py]: `return jsonify(message="...", id=node_id)`

File: backend/cdApp.py (hunk 1 above)
Comment (nfischer) [inline backend/cdApp.py]: Done.

File: backend/cdApp.py (hunk 2 above)
Comment (nfischer) [inline backend/cdApp.py]: Got it

File: backend/cdApp.py (hunk 3 above)
Comment (nfischer) [inline backend/cdApp.py]: I think @Sarabostani got this one working

File: backend/cdApp.py (hunk 4 above)
Comment (nfischer) [inline backend/cdApp.py]: Done

File: backend/cdApp.py (hunk 5 above)
Comment (nfischer) [inline backend/cdApp.py]: We should probably rename this to something more descriptive, like `node_tree`, which specifies what it represents

File: backend/cdApp.py (hunk 6 above)
Comment (nfischer) [inline backend/cdApp.py]: This API is written differently I think. Adding a child relationship means you specify the parent's ID in the data payload as well as the IDs of the children

File: backend/cdApp.py [hunk 7]
```diff
@@ -37,34 +37,121 @@ def dict_factory(cursor, row):
...
+            raise InvalidUsage('node_id is out of range')
+        return return_val
+
+class Children(Resource):
+    def put(self):
+        g.db.execute('INSERT INTO children (parent_id, children) values (?, ?)',
+                     [request.form['id'], request.form['children']])
         g.db.commit()
-        return jsonify(message='New node was successfully created')
+        return jsonify(message='Children were successfully added to the node', id='1')
```
Comment (sakekasi) [inline backend/cdApp.py]: still gotta resolve this issue. do the same as with the other one?

Comment (nfischer) [inline backend/cdApp.py]: Done. Should be correct, haven't tested though

File: backend/cdApp.py [hunk 8]
```diff
@@ -37,34 +37,121 @@ def dict_factory(cursor, row):
...
+        cursor = g.db.execute('''SELECT n.id, n.contents, n.renderer, c.children 
+                              FROM nodes AS n LEFT JOIN children AS c 
+                              ON n.id = c.parent_id WHERE id=%s''' % node_id)
+        return_val = cursor.fetchone()
+        if return_val is None:
+            raise InvalidUsage('node_id is out of range')
+        return return_val
+
+class Children(Resource):
+    def put(self):
```
Comment (sakekasi) [inline backend/cdApp.py]: can we make this self, nodeid instead of needing the id in the request?

Comment (sakekasi) [inline backend/cdApp.py]: `def put(self, node_id):`

Comment (nfischer) [inline backend/cdApp.py]: I'll take a look. I think it's pretty easy.

Comment (nfischer) [inline backend/cdApp.py]: This should be implemented now

**Comment by nfischer** [comment]
I think this is ready to merge. It definitely isn't perfect, but I think the API won't change so much (just aiming to improve error handling in the next few commits), and I think it's more important to focus on frontend.

**Pull request merged**